Source/Core/Core/IOS/USB/Emulated/Skylanders/SkylanderCrypto.cpp

```diff
@@ -0,0 +1,177 @@
+      0x46B4, 0xB75B, 0xA77A, 0x9719, 0x8738, 0xF7DF, 0xE7FE, 0xD79D, 0xC7BC, 0x48C4, 0x58E5,
+      0x6886, 0x78A7, 0x0840, 0x1861, 0x2802, 0x3823, 0xC9CC, 0xD9ED, 0xE98E, 0xF9AF, 0x8948,
+      0x9969, 0xA90A, 0xB92B, 0x5AF5, 0x4AD4, 0x7AB7, 0x6A96, 0x1A71, 0x0A50, 0x3A33, 0x2A12,
+      0xDBFD, 0xCBDC, 0xFBBF, 0xEB9E, 0x9B79, 0x8B58, 0xBB3B, 0xAB1A, 0x6CA6, 0x7C87, 0x4CE4,
+      0x5CC5, 0x2C22, 0x3C03, 0x0C60, 0x1C41, 0xEDAE, 0xFD8F, 0xCDEC, 0xDDCD, 0xAD2A, 0xBD0B,
+      0x8D68, 0x9D49, 0x7E97, 0x6EB6, 0x5ED5, 0x4EF4, 0x3E13, 0x2E32, 0x1E51, 0x0E70, 0xFF9F,
+      0xEFBE, 0xDFDD, 0xCFFC, 0xBF1B, 0xAF3A, 0x9F59, 0x8F78, 0x9188, 0x81A9, 0xB1CA, 0xA1EB,
+      0xD10C, 0xC12D, 0xF14E, 0xE16F, 0x1080, 0x00A1, 0x30C2, 0x20E3, 0x5004, 0x4025, 0x7046,
+      0x6067, 0x83B9, 0x9398, 0xA3FB, 0xB3DA, 0xC33D, 0xD31C, 0xE37F, 0xF35E, 0x02B1, 0x1290,
+      0x22F3, 0x32D2, 0x4235, 0x5214, 0x6277, 0x7256, 0xB5EA, 0xA5CB, 0x95A8, 0x8589, 0xF56E,
+      0xE54F, 0xD52C, 0xC50D, 0x34E2, 0x24C3, 0x14A0, 0x0481, 0x7466, 0x6447, 0x5424, 0x4405,
+      0xA7DB, 0xB7FA, 0x8799, 0x97B8, 0xE75F, 0xF77E, 0xC71D, 0xD73C, 0x26D3, 0x36F2, 0x0691,
+      0x16B0, 0x6657, 0x7676, 0x4615, 0x5634, 0xD94C, 0xC96D, 0xF90E, 0xE92F, 0x99C8, 0x89E9,
+      0xB98A, 0xA9AB, 0x5844, 0x4865, 0x7806, 0x6827, 0x18C0, 0x08E1, 0x3882, 0x28A3, 0xCB7D,
+      0xDB5C, 0xEB3F, 0xFB1E, 0x8BF9, 0x9BD8, 0xABBB, 0xBB9A, 0x4A75, 0x5A54, 0x6A37, 0x7A16,
+      0x0AF1, 0x1AD0, 0x2AB3, 0x3A92, 0xFD2E, 0xED0F, 0xDD6C, 0xCD4D, 0xBDAA, 0xAD8B, 0x9DE8,
+      0x8DC9, 0x7C26, 0x6C07, 0x5C64, 0x4C45, 0x3CA2, 0x2C83, 0x1CE0, 0x0CC1, 0xEF1F, 0xFF3E,
+      0xCF5D, 0xDF7C, 0xAF9B, 0xBFBA, 0x8FD9, 0x9FF8, 0x6E17, 0x7E36, 0x4E55, 0x5E74, 0x2E93,
+      0x3EB2, 0x0ED1, 0x1EF0};
+
+  u16 crc = init_value;
+
+  for (size_t i = 0; i < data.size(); ++i)
+  {
+    const u16 tmp = (crc >> 8) ^ data[i];
+    crc = (crc << 8) ^ CRC_CCITT_TABLE[tmp];
+  }
+
+  return crc;
+}
+// CRC-64 algorithm that is limited to 48 bits every iteration
+u64 ComputeCRC48(std::span<const u8> data)
+{
+  const u64 polynomial = 0x42f0e1eba9ea3693;
+  const u64 initial_register_value = 2ULL * 2ULL * 3ULL * 1103ULL * 12868356821ULL;
+
+  u64 crc = initial_register_value;
+  for (size_t i = 0; i < data.size(); ++i)
+  {
+    crc ^= (static_cast<u64>(data[i]) << 40);
+    for (size_t j = 0; j < 8; ++j)
+    {
+      if (Common::ExtractBit(crc, 47))
+      {
+        crc = (crc << 1) ^ polynomial;
+      }
+      else
+      {
+        crc <<= 1;
+      }
+    }
+  }
+  return crc & 0x0000FFFFFFFFFFFF;
+}
+u64 CalculateKeyA(u8 sector, std::span<const u8, 0x4> nuid)
+{
+  if (sector == 0)
+  {
+    return 73ULL * 2017ULL * 560381651ULL;
+  }
+
+  std::array<u8, 5> data = {nuid[0], nuid[1], nuid[2], nuid[3], sector};
+
+  u64 big_endian_crc = ComputeCRC48(data);
+  u64 little_endian_crc = Common::swap64(big_endian_crc) >> 16;
+
+  return little_endian_crc;
+}
+void ComputeChecksum(ChecksumType type, const u8* data_start, u8* output)
```

Comment:
The `ChecksumType` enum here seems a bit odd, this is never called with anything but a fixed value and the entire function is effectively just a big `switch (type)`. It might make more sense to just have every checksum type its own function, ie. `ComputeChecksumType0()` and so on. That way you can also make `data_start` a `std::span` with the correct known length of the data to be checksummed.

Reply:
I get what you mean, but I'm struggling with implementing it. Not only does converting an array to a span with length add one line of code per call to ComputeChecksumTypeX, some checksum types like type 1 also require a modification to the input, something that seemingly can't be done to a span.

Reply:
To be clear, I just mean that you have eg. `ComputeChecksumType1(std::span<const u8, 0x10> data, u8* output)` and then at the call site change `ComputeChecksumType1(decrypted.data() + area_offset, decrypted.data() + area_offset + 0xE);` to `ComputeChecksumType1(std::span<const u8, 0x10>(decrypted.data() + area_offset, 0x10), decrypted.data() + area_offset + 0xE);`. You can keep the copy into a local with modifications inside the `ComputeChecksumType1()` function. However, after trying it myself, I'm not sure if it really is all that better since you have to write the source size multiple times. So eh.

Just removing the enum and having individual functions per type however still looks better to me. You can take https://github.com/AdmiralCurtiss/dolphin/commit/5eed27fd7b4530b8d752c6aba7b15651cdb4fa19 for that if you want.

Reply:
Thanks. Just did